File: matchCalendars.py
def matchCalendars(calendar1, dailyBounds1, calendar2, dailyBounds2, duration):
    updatedCalendar1 = updateCalendar(calendar1, dailyBounds1)
    updatedCalendar2 = updateCalendar(calendar2, dailyBounds2)
    mergedCalendar = mergeCalendar(updatedCalendar1, updatedCalendar2)
    sortedCalendars = sortedCalendars(mergedCalendar)
    return getAvailabilities(sortedCalendars, duration)


def updateCalendar(calendar, dailyBounds):
    updatedCalendar = calendar[:]
    updatedCalendar.insert(0, ["0:00", dailyBounds[0]])
    updatedCalendar.append([dailyBounds[1], "23:59"])
    return list(map(lambda m: [timeToMinutes(m[0]), timeToMinutes(m[1])], updatedCalendar))


def mergeCalendar(calendar1, calendar2):
    merged = []
    i, j = 0, 0
    while i < len(calendar) and j < len(calendar2):
        meeting1, meeting2, = calendar1[i], calendar2[j]
        if meeting1[0] < meeting2[0]:
            merged.append(meeting)
            i += 1
        else:
            merged.append(meeting2)
            j += 1
    while i < len(calendar1):
        merged.append(calendar1[i])
        i += 1
    while j < len(calendar2):
        merged.append(calendar2[j])
        j += 1
    return merged


def sortedCalendars(calendar):
    sortedC = [calendar[0][:]]
    for i in range(1, len(calendar)):
        currentMeeting = calendar[i]
        previousMeeting = sortedC[-1]
        currentStart, currentEnd = currentMeeting
        previousStart, previousEnd = previousMeeting
        if previousEnd >= currentStart:
            newPreviousMeeting = [previousStart, max(previousEnd, currentEnd)]
            sortedC[-1] = newPreviousMeeting
        else:
            sortedC.append(currentMeeting[:])
    return sortedC


def getAvailabilities(calendar, duration):
    availabilities = []
    for i in range(1, len(calendar)):
        start = calendar[i -1][1]
        end = calendar[i][0]
        availableDuration = end - start
        if availableDuration >= duration:
            availabilities.append([start, end])
    return list(map(lambda m: [minutesToTime(m[0]), minutesToTime(m[1])], availabilities))


def timeToMinutes(time):
    hours, minutes = list(map(int, time.split(":")))
    return hours * 60 + minutes


def minutesToTime(minutes):
    hours = minutes // 60
    mins = minutes % 60
    hoursStr = str(hours)
    minutesStr = "0" + str(mins) if mins < 10 else str(mins)
    return hoursStr + ":" + minutesStr

File: matchCalendars.rs
fn main() {
    // Example usage
}

fn match_calendars(calendar1: Vec<Vec<&str>>, daily_bounds1: Vec<&str>, calendar2: Vec<Vec<&str>>, daily_bounds2: Vec<&str>, duration: i32) -> Vec<Vec<String>> {
    let updated_calendar1 = update_calendar(calendar1, daily_bounds1);
    let updated_calendar2 = update_calendar(calendar2, daily_bounds2);
    let merged_calendar = merge_calendar(updated_calendar1, updated_calendar2);
    let sorted_calendars = sort_calendars(merged_calendar);
    get_availabilities(sorted_calendars, duration)
}

fn update_calendar(calendar: Vec<Vec<&str>>, daily_bounds: Vec<&str>) -> Vec<Vec<i32>> {
    let mut updated = vec![vec!["0:00", daily_bounds[0]]];
    updated.extend_from_slice(&calendar);
    updated.push(vec![daily_bounds[1], "23:59"]);
    updated.iter().map(|m| vec![time_to_minutes(m[0]), time_to_minutes(m[1])]).collect()
}

fn merge_calendar(mut calendar1: Vec<Vec<i32>>, mut calendar2: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let mut merged: Vec<Vec<i32>> = Vec::new();
    while !calendar1.is_empty() && !calendar2.is_empty() {
        let meeting1 = calendar1[0].clone();
        let meeting2 = calendar2[0].clone();
        if meeting1[0] < meeting2[0] {
            merged.push(meeting1);
            calendar1.remove(0);
        } else {
            merged.push(meeting2);
            calendar2.remove(0);
        }
    }
    merged.extend(&calendar1);
    merged.extend(&calendar2);
    merged
}

fn sort_calendars(calendar: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let mut sorted = vec![calendar[0].clone()];
    for current in &calendar[1..] {
        let mut previous = sorted.last_mut().unwrap();
        if previous[1] >= current[0] {
            previous[1] = std::cmp::max(previous[1], current[1]);
        } else {
            sorted.push(current.clone());
        }
    }
    sorted
}

fn get_availabilities(calendar: Vec<Vec<i32>>, duration: i32) -> Vec<Vec<String>> {
    let mut availabilities: Vec<Vec<String>> = Vec::new();
    for i in 1..calendar.len() {
        let start = calendar[i - 1][1];
        let end = calendar[i][0];
        if end - start >= duration {
            availabilities.push(vec![minutes_to_time(start), minutes_to_time(end)]);
        }
    }
    availabilities
}

fn time_to_minutes(time: &str) -> i32 {
    let parts: Vec<&str> = time.split(':').collect();
    let hours: i32 = parts[0].parse().unwrap();
    let minutes: i32 = parts[1].parse().unwrap();
    hours * 60 + minutes
}

fn minutes_to_time(minutes: i32) -> String {
    let hours = minutes / 60;
    let mins = minutes % 60;
    format!("{}:{:02}", hours, mins)
}
